File: itx-java/itx-backend/src/main/java/io/github/chehsunliu/itx/backend/feature/user/UserRoutes.java
package io.github.chehsunliu.itx.backend.feature.user;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.chehsunliu.itx.backend.error.BackendException;
import io.github.chehsunliu.itx.backend.middleware.Envelope;
import io.github.chehsunliu.itx.backend.middleware.ItxContext;
import io.github.chehsunliu.itx.contract.repo.SubscriptionRepo;
import io.github.chehsunliu.itx.contract.repo.User;
import io.github.chehsunliu.itx.contract.repo.UserRepo;
import io.javalin.Javalin;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

public final class UserRoutes {
  private UserRoutes() {}

  public static void register(
      Javalin app, ObjectMapper mapper, UserRepo userRepo, SubscriptionRepo subscriptionRepo) {
    app.get(
        "/api/v1/users/me",
        ctx -> {
          ItxContext c = ItxContext.from(ctx);
          if (c.userEmail == null) throw BackendException.unknown("missing X-Itx-User-Email");
          User user = userRepo.upsert(new UserRepo.UpsertParams(c.userId, c.userEmail));
          Envelope.data(ctx, UserDto.fromUser(user));
        });

    app.get(
        "/api/v1/users/{id}/subscriptions",
        ctx -> {
          UUID id;
          try {
            id = UUID.fromString(ctx.pathParam("id"));
          } catch (IllegalArgumentException e) {
            throw BackendException.badRequest("invalid user id");
          }
          // Pre-check the subject so an unknown user yields 404, not an empty list.
          userRepo.get(id);
          List<User> authors = subscriptionRepo.listAuthors(id);
          List<UserDto> items =
              authors.stream().map(UserDto::fromUser).collect(Collectors.toList());
          Envelope.data(ctx, Map.of("items", items));
        });
  }
}

File: itx-kt/itx-backend/src/main/kotlin/io/github/chehsunliu/itx/backend/feature/user/UserRoutes.kt
package io.github.chehsunliu.itx.backend.feature.user

import io.github.chehsunliu.itx.backend.error.BackendException
import io.github.chehsunliu.itx.backend.feature.UuidSerializer
import io.github.chehsunliu.itx.backend.middleware.itxContext
import io.github.chehsunliu.itx.backend.middleware.respondData
import io.github.chehsunliu.itx.contract.repo.SubscriptionRepo
import io.github.chehsunliu.itx.contract.repo.User
import io.github.chehsunliu.itx.contract.repo.UserRepo
import io.ktor.server.application.call
import io.ktor.server.routing.Route
import io.ktor.server.routing.get
import io.ktor.server.routing.route
import io.ktor.server.util.getOrFail
import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class UserDto(
    @Serializable(with = UuidSerializer::class) val id: UUID,
    val email: String,
)

fun User.toDto(): UserDto = UserDto(id, email)

@Serializable data class ListSubscriptionsResponse(
    val items: List<UserDto>,
)

fun Route.userRoutes(
    userRepo: UserRepo,
    subscriptionRepo: SubscriptionRepo,
) {
    route("/users") {
        get("/me") {
            val ctx = call.itxContext()
            val email = ctx.userEmail ?: throw BackendException.unknown("missing X-Itx-User-Email")
            val user = userRepo.upsert(UserRepo.UpsertParams(ctx.userId!!, email))
            call.respondData(UserDto.serializer(), user.toDto())
        }

        get("/{id}/subscriptions") {
            val id = UUID.fromString(call.parameters.getOrFail("id"))
            // Pre-check the subject so an unknown user yields 404, not an empty list.
            userRepo.get(id)
            val authors = subscriptionRepo.listAuthors(id)
            call.respondData(
                ListSubscriptionsResponse.serializer(),
                ListSubscriptionsResponse(items = authors.map { it.toDto() }),
            )
        }
    }
}
